inst/app/modules/main_map.R

```diff
@@ -1,5 +1,23 @@
 
-hk_accidents_valid <- filter(hk_accidents, !is.na(latitude) & !is.na(longitude))
+# Get information of all types of vehicles involved in the accidents
+# TODO: prepare following mutated dataset before running shiny to save loading time?
+hk_vehicles_involved = hk_vehicles %>%
+  group_by(Serial_No_) %>%
+  summarize(vehicle_class_involved = paste(sort(unique(Vehicle_Class)), collapse = ", "))
+
+# Get information on whether the accidents involves pedestrian and number of pedestrians
+hk_casualties_pex = hk_casualties %>%
+  group_by(Serial_No_) %>%
+  summarise(
+    include_pex = any(Role_of_Casualty == "Pedestrian"),
+    n_pex_involved = sum(Role_of_Casualty == "Pedestrian")
+  )
+
+hk_accidents_join <- hk_accidents %>%
+  left_join(hk_casualties_pex, by = "Serial_No_") %>%
+  left_join(hk_vehicles_involved, by = "Serial_No_")
```

Comment:
The shiny app now has to run these intensive data manipulation steps for preparing the texts shown in map popups. That's why the collision map viewer now takes ~10 second (or more) to initialise. Any idea how to improve the performance?

Reply:
Doesn't look like this is dynamic/dependent on user-input so you have two options to make it faster:
1. Shift the data manipulation to `global.R` by adding this module in there and then calling it in `server.R` - this will increase initialisation times at the start.
2. Manipulate the data before it goes in `global.R` so you're just loading the manipulated dataset - this is probably the best solution and it does not increase initialisation nor server wait times.

What do you think?

Reply:
I also thought of completing all data manipulation steps and export the output as a CSV file. Then we could load it directly. But I have two concerns with this approach:

1. Where should we store the manipulated dataset in the project? A new folder in `/inst/app`?
2. Where should we store the codes used for data manipulation? Or even should these codes be included in the repo?

Reply:
Good questions:

1. Yeah, I think it is fine to store this in a new folder in the repo if the file size is not large. Alternatively, could store it on Google Drive and import the data file from there.
2. The data manipulation code should sit in the repo because it is part of the work for this. I'd store it outside the `/inst/` subdirectory and probably have a `src/` folder in the top-level directory for this project where this script will go - however, feel free to have a look at best practice to see how other Shiny app repos deal with something like this.

Reply:
Thanks for the suggestions! I exported the manipulated dataset and the size is about 23 MB. The size could vary, depending on the number of filtering widgets we would like to add.

Maybe we could refactor this part in a new PR?

Reply:
Agreed, could you raise a GitHub Issue off the back of this so we can track this?

Reply:
Now in #13 for further discussions.